### ISSUE TITLE
Fail to upload big file with mc : need multipart upload support

### ISSUE THREAD

**Issue by zaibon** [gigforks/minio#32]
```
mc: <ERROR> Failed to copy `/home/zaibon/videos/GOPR2668.MP4`. A header you provided implies functionality that is not implemented
mc: <ERROR> Session safely terminated. To resume session `mc session resume GyDBpnkr`
```

**Comment by iwanbk** [comment]
It seems you wrote a quite big file, right?

I tested with 1.3Gb of file, it tries to use multipart upload, which indeed not supported according to the spec.
Thinking of this, i think multipart upload should be implemented on gateway because 0-stor already has mechanism to upload big file

**Comment by zaibon** [comment]
It is indeed need to be implement on the gateway, this had nothing to do with the backend.

**Comment by iwanbk** [comment]
> this had nothing to do with the backend.

well,  bacblaze and aliyun gateway doesn't handle this, already handled by the respective libraries.
Not sure about the other.

It shouldn't be hard if the client only talk to single minio server. (thinking of distributed minio).

**Comment by iwanbk** [comment]
In short, this is how multipart upload work:
- client initiate multipart upload, then the server return multipart upload ID (uploadID)
- client send all the parts, one by one, identified by part id. It doesn't necessarily in consecutive order.
- client end the upload process by sending `complete` or `abort` command.
- in case of `complete` command, server creates an object by concatenating the parts in ascending order based on the part number

For this, we need a kind of buffer or temporary storage to store the parts.

I can see two options to implement the temporary storage:
(1) as temporary object, in 0-stor
pros:
- reliable, we shouldn't  loss the parts  in case of crash
cons:
- slow when doing `complete`, because we need to read from server and then write again

(2) as temporary file in minio server
pros:
- faster
cons:
- might loss the parts in case of crash -> i don't think it is a great deal

I prefer the 2nd solution, because i don't think that it is a big deal to lost the uploaded parts. client will always know that he is losing the parts when he continue the operation, either uploading next part or completing the operation.

**Comment by iwanbk** [comment]
We go to with option no 1 for these reasons:
- minio instance won't have storage
- option (2) won't work in distributed minio. Our current filemeta won't work either, but this is a different issue

**Comment by iwanbk** [comment]
happy path for multipart upload is working on `multipart` branch.
( @zaibon , in case you want to test it).

It still need to be:
- cleaned up
- add tests
- add `abort`

**Comment by iwanbk** [comment]
Fixed in #38, tested with unit test and uploading 1.6GB of file
